src/server/transaction.h

```diff
@@ -433,6 +429,24 @@ class Transaction {
     }
   };
 
+  // Barrier akin to helio's BlockingCounter, but with phase counting and proper acquire semantics
+  // for polling work from other threads. And without keap allocation.
+  struct PhasedBarrier {
+    uint32_t Start(uint32_t count);  // Release: Store count, return phase
+    bool Active() const;             // Acquire: Check if count > 0. Use for polling for work
+
+    void Dec(Transaction* keep_alive);  // Relase: Decrease count, notify ec on 0
+    void Wait();                        // Acquire: Wait until count = 0, increment phase
+
+    bool IsMyPhase(uint32_t phase) const;  // Assumes already Acquired. Return true if same phase
+    uint32_t DEBUG_Get() const;
+
+   private:
+    std::atomic_uint32_t phase_{0};
+    std::atomic_uint32_t count_{0};
+    EventCount ec_{};
+  };
+
```

Comment:
My take is that extracting something into a simple abstraction (but with a clearly defined interface) makes the code much easier to reason about, because we now think in terms of the 4 operations, instead of raw atomics and events, mixed with regular code

Reply:
It's sad we can't reuse helio's BlockingCounter for this. We could, but we have to get rid of heap allocation and allow peeking with acquire semantics... Which is more complicated than counting manually